doc/cursors.rst

```diff
@@ -3,16 +3,435 @@
 Cursors
 =======
 
-.. admonition:: Work in progress
-   :class: danger
+Built upon :ref:`lenses`, cursors are a great way to bridge
+value-oriented designs and object-oriented systems. It also allows
+modularity.
 
-   This section of the documentation has not been written yet.
+.. _types-of-cursors:
 
-   Meanwhile, you can watch the **C++ Russia 2019 Talk**: `Squaring
-   the circle: value oriented design in an object oriented system
-   <https://www.youtube.com/watch?v=e2-FRFEx8CA>`_ (`slides`_).
+Types of cursors
+----------------
 
-.. _slides: https://sinusoid.es/talks/cpprussia19-piter
+Cursors are wrappers for :ref:`models <model>` where you can get the current
+version of a model, watch it for changes, zoom into other cursors
+and (for some types of cursors) replace the model with another version:
+
+.. code-block:: c++
+
+   // Call signatures of cursors
+   model_type cursor_type<model_type>::get();
+   void lager::watch(cursor_type<model_type> cursor,
+                     std::function<void(model_type)> callback);
+   // for some types only
+   void cursor_type<model_type>::set(model_type new_model);
+
+Lager provides four types of cursors:
```

Comment:
There are some more cursors that are not mentioned here: `lager::writer`, `lager::sensor`, `lager::constant` and "lens cursros" and "transduced cursor".

I like distinguishing between two types of cursors, "interface" cursors, "root" cursors, and "derived" cursors. `reader`, `writer` and `cursors` are the generic, interface, cursors. All other cursors are convertible to it.

Then we have "root" cursors, that are sources of data, roots of the cursor tree. These are `state`, `store`, `sensor` and `constant`.

Then we have "derived" cursors, that are created with `with(...)`, `.xform()`, `.zoom()` and `operator[]` from other cursors, forming inner and leaf nodes of the cursor tree.